### PR TITLE
Add tests ensuring that cudf's default stream is always used

### LINKED ISSUE #11929: [BUG] multibyte_split implicitly synchronizes all streams
**Describe the bug**
Currently `multibyte_split` uses a stream pool to divide up work between multiple streams. However, `output_builder.inplace_resize` still resizes an `rmm::device_uvector` on [CUDA's default stream](https://github.com/rapidsai/cudf/blob/branch-22.12/cpp/src/io/text/multibyte_split.cu#L528). This function is being called from `output_builder.advance_output`, which does not currently accept a stream. This is problematic since the default stream implicitly synchronizes with all streams. 

**Expected behavior**
From a quick analysis it is not clear to me whether `advance_output` or `inplace_resize` should be running on the `scan_stream` or the `read_stream` inside `multibyte_split`. [Based on where `advance_output` is called](https://github.com/rapidsai/cudf/blob/branch-22.12/cpp/src/io/text/multibyte_split.cu#L553k), I suspect that it actually requires an explicit synchronization between these two streams, and perhaps the existing events are already intended to solve that need. Fixing this would require modifying this function to run on the appropriate stream, and perhaps introducing an explicit synchronization or some appropriate event waiting.

### LINKED ISSUE #11942: [FEA] Ensure that all of cudf's algorithms run on cudf's default stream by default
**Is your feature request related to a problem? Please describe.**
libcudf currently runs entirely on CUDA's default stream. Internally, this is managing using an `rmm::cuda_stream_view` stored as a global variable used everywhere. We would like to move towards enabling stream-ordered libcudf APIs. In order to enable this safely, we need to ensure that libcudf is properly using its default stream everywhere in the codebase.

**Describe the solution you'd like**
We should write a separate library that overrides existing CUDA functions (via dlsym) to throw an error if they are invoked using CUDA's default stream. We should then preload this library when running libcudf's test suite so that our unit tests can verify that we are properly passing `cudf::default_stream_value` through to everywhere that it needs to be in the code. We will need to make cudf's default stream configurable so that it can be set to a non-default value when running the tests.

**Describe alternatives you've considered**
None

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This PR ensures that cudf's default stream is properly passed to all kernel launches so that nothing implicitly runs on the CUDA default stream. It adds a small library that is built during the tests and overloads CUDA functions to throw an exception when usage of the default stream is detected. It also fixes all remaining usage of anything other than cudf's default stream (I fixed most of the issues in previous PRs, but I found a few others when finalizing this one).

Resolves #11929
Resolves #11942 

### Important notes for reviewers:
- **The changeset is deceptively large.** The vast majority of the changes are just a global find-and-replace of `cudf::get_default_stream()` for `cudf::default_stream_value`, as well as a few smaller fixes such as missing `CUDF_TEST_PROGRAM_MAIN` in a couple of tests and usage of `rmm::cuda_stream_default`. The meaningful changes are:
    - The new default stream getter/setter in `default_stream.[hpp|cpp]`
    - The addition of `cpp/tests/utilities/identify_stream_usage`
    - The changes to the base testing fixture in `cpp/include/cudf_test/base_fixture.hpp` to inject the custom stream.
    - The changes to CI in `ci/gpu/build.sh` to build and use the new library.
- This PR is a breaking change because it moves the default stream into the detail namespace. Going forward the default stream may only be accessed using the public accessor `cudf::get_default_stream()`. I have added a corresponding setter, but it is also in the detail namespace since I do not want to publicly support changing the default stream yet, only for the purpose of testing. Reviewers, please leave comments if you disagree with those choices.
- I have made getting and setting the default stream thread-safe, but there is still only a single stream. In multi-threaded applications we may want to support a stream per thread so that users could manually achieve PTDS with more fine-tuned control. Is this worthwhile? Even if it is, I'm inclined to wait for a subsequent PR to implement this unless someone feels strongly otherwise.
- I'm currently only overloading `cudaLaunchKernel`. I can add overloads for other functions as well, but I didn't want to go through the effort of overloading every possible API. If reviewers have a minimal set that they'd like to see overloaded, let me know. [I've included links to all the relevant pages of the CUDA runtime API in the identify_stream_usage.cu file](https://github.com/rapidsai/cudf/pull/11875/files#diff-0b2762207c27c080acd2114475c7a1c06377a7c18c4e9c3de60ecbdc82a4dc61R99) if someone wants to look through them.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cu
Comment (davidwendt) [inline cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cu:130]: Could this just check if `stream != cudf::get_default_stream().value()` ?

**Review by davidwendt — COMMENTED** [review]
Does `identify_stream_usage.cu` need to be a `.cu` file? There is no device here for nvcc to compile.

File: cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cu
Comment (vyasr) [reply cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cu:130]: It could, and I originally considered that. However, there are two issues:
- That won't necessarily capture usage of `thrust::device` because by default `cudf::default_stream_value` is 0 while (based on my observations) `thrust::device` is `cudaStreamLegacy`(==1).
- That complicates the build because now we need to link this library against libcudf. It's possible, but it feels rather circular to link this against libcudf only to preload it before the run of libcudf. Also, within the context of a CI run where we install conda packages at some point I figured it would make maintenance easier to avoid `find_package(cudf)`.

**Comment by vyasr** [comment]
> Does `identify_stream_usage.cu` need to be a `.cu` file? There is no device here for nvcc to compile.

I think it could if we wanted to include the necessary headers. I can remove the `__host__` attributes and leave that implicit, but I would still need to be able to see the types `cudaError_t` and `cudaStream_t`. Do you think that it's cleaner to have a cpp file that includes `driver_types.h` (or whichever other headers that we need)? I'm not sure that it's any better than just compiling directly with nvcc.

File: cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cu
Comment (davidwendt) [reply cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cu:130]: I figured it was something like this. Regardless it may be worth documenting this in the source file for future generations.

**Comment by davidwendt** [comment]
> > Does `identify_stream_usage.cu` need to be a `.cu` file? There is no device here for nvcc to compile.
> 
> I think it could if we wanted to include the necessary headers. I can remove the `__host__` attributes and leave that implicit, but I would still need to be able to see the types `cudaError_t` and `cudaStream_t`. Do you think that it's cleaner to have a cpp file that includes `driver_types.h` (or whichever other headers that we need)? I'm not sure that it's any better than just compiling directly with nvcc.

I do think it should be a `.cpp` file if it can. The only header you should need is `cuda_runtime.h`

File: cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cu
Comment (vyasr) [reply cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cu]: Done.

**Comment by vyasr** [comment]
> > > Does `identify_stream_usage.cu` need to be a `.cu` file? There is no device here for nvcc to compile.
> > 
> > 
> > I think it could if we wanted to include the necessary headers. I can remove the `__host__` attributes and leave that implicit, but I would still need to be able to see the types `cudaError_t` and `cudaStream_t`. Do you think that it's cleaner to have a cpp file that includes `driver_types.h` (or whichever other headers that we need)? I'm not sure that it's any better than just compiling directly with nvcc.
> 
> I do think it should be a `.cpp` file if it can. The only header you should need is `cuda_runtime.h`

Works for me. I've made that change now.

**Comment by jrhemstad** [comment]
> * I'm currently only overloading `cudaLaunchKernel`

I don't think this is sufficient.  At minimum you need any memcpy APIs. Probably events too.

**Comment by vyasr** [comment]
> > * I'm currently only overloading `cudaLaunchKernel`
> 
> I don't think this is sufficient. At minimum you need any memcpy APIs. Probably events too.
> 
> Oh, and `cudaMalloc/FreeAsync`.

I considered these but wasn't sure of the scope since we rely on rmm to pretty much handle all of that for us. But I guess we may as well include those here, then if there are issues that need upstream fixes we can do that.

Why events? Is it because nvtx uses those, or do we have internal uses of them? Happy to add the overloads of the appropriate APIs.

> Part of this test suite should be wrapping the `device_memory_resource` with an adaptor that performs the same checks as your LD_PRELOAD overloads to ensure all calls to `device_memory_resource::allocate` are done with the expected stream.

Wouldn't it be better to just rely on checking the underlying CUDA APIs? We aren't guaranteed that all types of memory resources are stream-ordered, so these tests wouldn't pass if the `cuda_memory_resource` was used, for instance (since that uses `cudaMalloc`). Checking at the memory resource level seems like it adds an extra layer without much benefit. What am I missing?

**Review by jlowe — APPROVED** [review]
Java approval

**Comment by jrhemstad** [comment]
> I considered these but wasn't sure of the scope since we rely on rmm to pretty much handle all of that for us. But I guess we may as well include those here, then if there are issues that need upstream fixes we can do that.

RMM is garbage-in, garbage out. If libcudf provides an incorrect stream then RMM will happily just use it. There's zero checking that RMM will do for you on this front. 

Also, there are plenty of uses of `cudaMemcpyAsync` in libcudf outside of RMM. 

> Why events? Is it because nvtx uses those, or do we have internal uses of them? Happy to add the overloads of the appropriate APIs.

RMM is the primary use of event APIs today, but it would be entirely possible for Thrust or CUB functions to use events (they don't, but they could!) or for future libcudf functionality to use events as well. 

> Wouldn't it be better to just rely on checking the underlying CUDA APIs?

Maybe, but only if we significantly expand our coverage of CUDA APIs. A memory resource implementation may never touch one of the currently covered CUDA APIs (they _definitely_ won't touch touch `cudaLaunchKernel`).

Furthermore, I'd bet the farm that `device_memory_resource::allocate` is the most commonly invoked function in libcudf that accepts a stream. So it gives you a wide net to catch misuses of streams. 

It's also pretty trivial to build such a custom resource adapter. Far easier than the LD_PRELOAD stuff you're doing. So it's just another piece to add to a multi-layered strategy.

**Comment by vyasr** [comment]
OK I've added overloads for all the CUDA APIs that I think could possibly be important. [There is a long-form comment in the code with links to all the CUDA API pages that I thought could be relevant](https://github.com/rapidsai/cudf/pull/11875/files#diff-e534d46ee31368763c4ec1770af0e5ba2a3790bd22be8cf04b94e032625388a1R154). The ones where I have overloaded APIs are marked as "Done". @jrhemstad @davidwendt please let me know if you see any APIs that I may have missed.

I would welcome feedback from anyone with strong macro-fu than I have; my current macro isn't terrible, but it is more repetitive than I would like between the signature and the arguments.

I'm going to work on adding tests as well as an rmm adapter like Jake suggested, then I think this will be good to go pending reviews.

**Comment by vyasr** [comment]
I have now added a direct test of this functionality for `cudaLaunchKernel` that builds and runs as part of CI before the tool is used. Adding new overloads is also fairly straightforward now (if slightly more verbose than I would like; again, would welcome feedback from anyone more skilled with macros). I also added a new rmm resource adapter that is used whenever we run on a custom stream. I think I've addressed all requests from everyone, this should be ready for another round of review.

**Comment by robertmaynard** [comment]
My first idea was to simplify the argument handling but that can't done since you need to always have the `cudaStream_t` parameter to alway have the name `stream` ( and usable ).

It does look like we have a consistent return type across all functions and the passing of `__host__` and `__device__` seems like complexity that we can save by not having a shared function. So something like this would be my suggestion https://godbolt.org/z/4dr5xvh6v

**Review by robertmaynard — CHANGES_REQUESTED** [review]
I don't think we should be offering a `__device__` overload for these functions since we can't track them

**Comment by vyasr** [comment]
> I don't think we should be offering a `__device__` overload for these functions since we can't track them

That is a great point that I thought about early on then apparently forgot somewhere in the process. However, I am not quite sure how to do this. Attributes are not sufficient to create a new overload right, so if I define these functions as just `__host__` won't that still wipe out the original `__host__` `__device__` definitions provided by the CTK? What we really want is to override the `__host__` functions but leave the device functions intact. Is that possible?

**Comment by robertmaynard** [comment]
> > I don't think we should be offering a `__device__` overload for these functions since we can't track them
> 
> That is a great point that I thought about early on then apparently forgot somewhere in the process. However, I am not quite sure how to do this. Attributes are not sufficient to create a new overload right, so if I define these functions as just `__host__` won't that still wipe out the original `__host__` `__device__` definitions provided by the CTK? What we really want is to override the `__host__` functions but leave the device functions intact. Is that possible?

Good points. I don't know how we can safely overload any function with __device__ markup since attributes aren't sufficent for new overloads with nvcc. Even if we try to define a `__host__` only version it doesn't matter since the cuda_runtime header will have already been included and marked the function as host and device.

I think what we want to do is compile this class as a c++ file and not CUDA. Drop both the __host__ and __device__ markups. That will ensure we only evaluate the functions for 'host'.

**Comment by vyasr** [comment]
It's already a cpp file compiled with gcc, but we need the cuda runtime header so that we can use the various cuda types like cudaStream_t etc. Are we already doing what we need to here, then? If I just drop the `__host__ __device__` specifiers do you think that will be sufficient, or is simply including cuda_runtime.h already enough to cause problems?

**Comment by robertmaynard** [comment]
> It's already a cpp file compiled with gcc, but we need the cuda runtime header so that we can use the various cuda types like cudaStream_t etc. Are we already doing what we need to here, then? If I just drop the `__host__ __device__` specifiers do you think that will be sufficient, or is simply including cuda_runtime.h already enough to cause problems?

I missed that it already was a .cpp file and was thinking it was a .cu. Yes just drop the `__host__` and `__device__` markup and we will be good ( no worries about including cuda_runtime.h).

File: cpp/include/cudf/utilities/default_stream.hpp
Comment (bdice) [inline cpp/include/cudf/utilities/default_stream.hpp:28]: Yup, this structure is fine. Move to another header as needed.

**Comment by vyasr** [comment]
Currently blocked by https://github.com/rapidsai/cudf/issues/11929

File: cpp/include/cudf/utilities/default_stream.hpp
Comment (vyasr) [reply cpp/include/cudf/utilities/default_stream.hpp:48]: Sounds good I'll leave everything here as is for now, then.

File: cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cpp
Comment (vyasr) [reply cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cpp:41]: That is a good question. My godbolt investigations indicated that the answer is no, at least for clang, and since we don't actually support any other compilers for cudf at this point it probably isn't worth thinking about this too much. I only included this guard because I saw that the demangling functions are implementation-defined and I have not tested any other compilers.

File: cpp/src/io/text/multibyte_split.cu
Comment (upsj) [inline cpp/src/io/text/multibyte_split.cu:535]: this change was probably not intended

File: cpp/src/io/text/multibyte_split.cu
Comment (upsj) [inline cpp/src/io/text/multibyte_split.cu]: I find it a bit surprising that one member call uses `scan_stream` and the other one uses `stream`. Do you see any harm in using `scan_stream` everywhere? If you're worried about the stream pools "leaking", we could introduce another scope between fork_stream and join_stream?

**Review by upsj — COMMENTED** [review]
Reviewing only the multibyte_split changes: LGTM

File: cpp/src/io/text/multibyte_split.cu
Comment (vyasr) [reply cpp/src/io/text/multibyte_split.cu:535]: Ah thanks.

File: cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cpp
Comment (bdice) [inline cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cpp:143]: Can we have a comment to explain what this line does?

**Review by bdice — APPROVED** [review]
I approve the overall design, though there are some elements that are unfamiliar to me, such as the library preloading and similar "injection tricks" to change the default stream for testing. This may deserve additional inspection by other reviewers.

**Review by sevagh — APPROVED** [review]

File: ci/gpu/build.sh
Comment (vyasr) [reply ci/gpu/build.sh:218]: I agree. My thinking is that the approach in this PR is a bit of a stopgap until I can implement #11943 (which will take significantly longer since it depends on the API changes to enable streams). At that point we will probably want to implement a more surgical approach to skipping the check for specific tests or determining more precisely what streams to allow or prevent usage of.

**Review by jrhemstad — CHANGES_REQUESTED** [review]
I have significant concerns about adding acquiring a mutex at so many places throughout the code just to enable this test case. Vyas and I will be meeting to discuss alternative solutions, but requesting changes for now to prevent merging as I do not think the pros outweigh the cons in the current solution.

File: ci/gpu/build.sh
Comment (vyasr) [reply ci/gpu/build.sh:218]: @davidwendt I chatted with @jrhemstad earlier today and he pointed out that this may become a moot point in the not too distant future since [spans will soon be in libcudacxx](https://github.com/NVIDIA/libcudacxx/pull/313) and we will eventually want to drop our spans in favor of those. That being the case, I've made a note in that test as you requested, but I don't think it's worth the effort to do anything else.

**Comment by vyasr** [comment]
> I have significant concerns about adding acquiring a mutex at so many places throughout the code just to enable this test case. Vyas and I will be meeting to discuss alternative solutions, but requesting changes for now to prevent merging as I do not think the pros outweigh the cons in the current solution.

I've resolved this issue by instead overloading `get_default_stream` in the `identify_stream_usage` library with a version that creates a function local static. The resulting stream is initialized after `main` so that it does not trigger the CUDA UB that I was running into in earlier iterations, and it obviates the need for the `set_default_stream` function, which has now been removed. Since the default stream is no longer settable, the mutex is no longer necessary.

**Review by jrhemstad — APPROVED** [review]

**Review by davidwendt — APPROVED** [review]

**Comment by vyasr** [comment]
I  just merged branch-22.12 locally, rebuilt, and reran the tests to make sure that nothing has been merged since the last test run that would use a default stream and break this, so I think this is ready to merge and shouldn't cause any test failures afterwards.

@gpucibot merge

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**
